File: truth_set/javascript/visual-representation/grouping/grouping.js
class ProductService {
    constructor() {
        this.apiUrl = 'https://api.example.com';
        this.products = [];
    }

    createProduct(name, price) {
        const newProduct = {
            id: this.generateId(),
            name,
            price,
            createdAt: new Date(),
        };
        this.products.push(newProduct);
        console.log(`Product "${name}" created with ID ${newProduct.id}.`);

        fetch(`${this.apiUrl}/products`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(newProduct),
        })
            .then(response => response.json())
            .then(data => console.log('Product saved to API:', data))
            .catch(error => console.error('Error saving product to API:', error));

        return newProduct;
    }

    deleteFlag = false;

    deleteProduct(id) {
        const index = this.products.findIndex(p => p.id === id);
        if (index !== -1) {
            const deletedProduct = this.products.splice(index, 1)[0];
            console.log(`Product "${deletedProduct.name}" with ID ${id} deleted.`);
        } else {
            console.log(`Product with ID ${id} not found.`);
        }
    }

    updateFlag = false;

    async updatePrice(id, newPrice) {
        try {
            const response = await fetch(`${this.apiUrl}/products/${id}`);
            const product = await response.json();
            product.price = newPrice;
            this.updateFlag = true;
            console.log(`Product ID ${id} price updated to $${newPrice}.`);

            fetch(`${this.apiUrl}/products/${id}`, {
                method: 'PUT',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify(product),
            })
                .then(response => response.json())
                .then(data => console.log('Product updated in API:', data))
                .catch(error => console.error('Error updating product in API:', error));

            return product;
        } catch (error) {
            console.error('Error fetching product from API:', error);
            return undefined;
        }
    }

    logProductDetails(id) {
        const product = this.products.find(p => p.id === id);
        if (product) {
            console.log(`ID: ${product.id}, Name: ${product.name}, Price: $${product.price}, Created At: ${product.createdAt}`);
        } else {
            console.log(`Product with ID ${id} not found.`);
        }
    }

    getAllProducts() {
        fetch(`${this.apiUrl}/products`)
            .then(response => response.json())
            .then(data => {
                this.products = data;
                console.log('Products fetched from API:', data);
            })
            .catch(error => console.error('Error fetching products from API:', error));

        return this.products;
    }

    status = 'active';

    generateId() {
        return this.products.length > 0 ? Math.max(...this.products.map(p => p.id)) + 1 : 1;
    }
}

const productService = new ProductService();

productService.createProduct('Tablet', 300);
productService.updatePrice(1, 350);
productService.logProductDetails(1);
productService.deleteProduct(1);
console.log(productService.getAllProducts());
\ No newline at end of file

File: truth_set/python/visual-representation/grouping/grouping.py
from datetime import datetime
import requests

class ProductService:
    def __init__(self):
        self.api_url = 'https://api.example.com'
        self.products = []

    def create_product(self, name, price):
        new_product = {
            'id': self.generate_id(),
            'name': name,
            'price': price,
            'created_at': datetime.now()
        }
        self.products.append(new_product)
        print(f'Product "{name}" created with ID {new_product["id"]}.')

        self._save_to_api(new_product)
        return new_product

    status = 'active'

    def delete_product(self, id):
        index = next((i for i, p in enumerate(self.products) if p['id'] == id), -1)
        if index != -1:
            deleted_product = self.products.pop(index)
            print(f'Product "{deleted_product["name"]}" with ID {id} deleted.')
        else:
            print(f'Product with ID {id} not found.')

    delete_flag = False

    def update_flag_status(self):
        self.update_flag = True

    update_flag = False

    async def update_price(self, id, new_price):
        try:
            product = await self._fetch_product(id)
            product['price'] = new_price
            self.update_flag = True
            print(f'Product ID {id} price updated to ${new_price}.')

            self._update_api(id, product)
            return product
        except Exception as error:
            print('Error fetching product from API:', error)
            return None

    def log_product_details(self, id):
        product = next((p for p in self.products if p['id'] == id), None)
        if product:
            print(f'ID: {product["id"]}, Name: {product["name"]}, Price: ${product["price"]}, Created At: {product["created_at"]}')
        else:
            print(f'Product with ID {id} not found.')

    def get_all_products(self):
        self._fetch_products()
        return self.products

    def generate_id(self):
        return max([p['id'] for p in self.products], default=0) + 1

    def _save_to_api(self, product):
        response = requests.post(
            f'{self.api_url}/products',
            json=product,
            headers={'Content-Type': 'application/json'}
        )
        if response.ok:
            print('Product saved to API:', response.json())
        else:
            print('Error saving product to API:', response.text)

    def _update_api(self, id, product):
        response = requests.put(
            f'{self.api_url}/products/{id}',
            json=product,
            headers={'Content-Type': 'application/json'}
        )
        if response.ok:
            print('Product updated in API:', response.json())
        else:
            print('Error updating product in API:', response.text)

    def _fetch_product(self, id):
        response = requests.get(f'{self.api_url}/products/{id}')
        if response.ok:
            return response.json()
        raise Exception(response.text)

    def _fetch_products(self):
        response = requests.get(f'{self.api_url}/products')
        if response.ok:
            self.products = response.json()
            print('Products fetched from API:', self.products)
        else:
            print('Error fetching products from API:', response.text)

product_service = ProductService()

product_service.create_product('Tablet', 300)
product_service.update_price(1, 350)
product_service.log_product_details(1)
product_service.delete_product(1)
print(product_service.get_all_products())
\ No newline at end of file